src/sv2/connman.h

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2023-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SV2_CONNMAN_H
+#define BITCOIN_SV2_CONNMAN_H
+
+#include <common/sockman.h>
+#include <sv2/messages.h>
+#include <sv2/transport.h>
+#include <pubkey.h>
+
+namespace {
+    /*
+     * Supported Stratum v2 subprotocols
+     */
+    static constexpr uint8_t TP_SUBPROTOCOL{0x02};
+
+    static const std::map<uint8_t, std::string> SV2_PROTOCOL_NAMES{
+    {0x02, "Template Provider"},
+    };
+}
+
+struct Sv2Client
+{
+    /* Ephemeral identifier */
+    size_t m_id;
+
+    /**
+     * Transport
+     */
+    std::unique_ptr<Sv2Transport> m_transport;
+
+    /**
+     * Whether the client has confirmed the connection with a successful SetupConnection.
+     */
+    bool m_setup_connection_confirmed = false;
+
+    /**
+     * Whether the client is a candidate for disconnection.
+     */
+    bool m_disconnect_flag = false;
+
+    /** Queue of messages to be sent */
+    std::deque<Sv2NetMsg> m_send_messages;
+
+    /**
+     * Whether the client has received CoinbaseOutputConstraints message.
+     */
+    bool m_coinbase_output_constraints_recv = false;
+
+    /**
+     * Specific additional coinbase tx output size required for the client.
+     */
+    unsigned int m_coinbase_tx_outputs_size;
+
+    explicit Sv2Client(size_t id, std::unique_ptr<Sv2Transport> transport) :
+                       m_id{id}, m_transport{std::move(transport)} {};
+
+    bool IsFullyConnected()
+    {
+        return !m_disconnect_flag && m_setup_connection_confirmed;
+    }
+
+    Sv2Client(Sv2Client&) = delete;
+    Sv2Client& operator=(const Sv2Client&) = delete;
+};
+
+/**
+ * Interface for sv2 message handling
+ */
+class Sv2EventsInterface
+{
+public:
+    virtual ~Sv2EventsInterface() = default;
+};
+
+/*
+ * Handle Stratum v2 connections.
+ * Currently only supports inbound connections.
+ */
+class Sv2Connman : SockMan
+{
+private:
+    /** Interface to pass events up */
+    Sv2EventsInterface* m_msgproc;
+
+    /**
+     * The current protocol version of stratum v2 supported by the server. Not to be confused
+     * with byte value of identitying the stratum v2 subprotocol.
+     */
+    const uint16_t m_protocol_version = 2;
+
+    /**
+     * The currently supported optional features.
+     */
+    const uint16_t m_optional_features = 0;
+
+    /**
+     * The subprotocol used in setup connection messages.
+     * An Sv2Connman only recognizes its own subprotocol.
+     */
+    const uint8_t m_subprotocol;
+
+    CKey m_static_key;
+
+    XOnlyPubKey m_authority_pubkey;
+
+    std::optional<Sv2SignatureNoiseMessage> m_certificate;
+
+    /**
+     * A map of all connected stratum v2 clients.
+     */
+    using Clients = std::unordered_map<NodeId, std::shared_ptr<Sv2Client>>;
+    Clients m_sv2_clients GUARDED_BY(m_clients_mutex);
```

Comment:
Why does this need a lock? I'm curious because in HTTP I think all client-related operations take place exclusively in the Sockman I/O thread.

Reply:
I think because of `DisconnectFlagged()`, `GetClientById()` or `ForEachClient()` - or for the test-only `(Fully)ConnectedClients()`. But it could be a left-over of earlier versions.

Reply:
In HTTP you probably don't have a need to "proactively" talk to a specific client.

Reply:
TSan will tell you